internal/serialization/default_compact_reader.go

```diff
@@ -0,0 +1,146 @@
+	offset := input.ReadByteAtPosition(variableOffsetsPos + index)
+	if offset == 0xFF {
+		return NULL_OFFSET
+	}
+	return int32(offset)
+}
+
+type DefaultCompactReader struct {
+	offsetReader            OffsetReader
+	in                      *ObjectDataInput
+	serializer              CompactStreamSerializer
+	schema                  Schema
+	dataStartPosition       int32
+	variableOffsetsPosition int32
+}
+
+func (r DefaultCompactReader) ReadInt32(fieldName string) int32 {
+	fd := r.getFieldDefinition(fieldName)
+	fieldKind := fd.fieldKind
+	switch fieldKind {
+	case pserialization.FieldKindInt32:
+		position := r.readFixedSizePosition(fd)
+		return r.in.ReadInt32AtPosition(position)
+	default:
+		panic(r.unexpectedFieldKind(fieldKind, fieldName))
+	}
+}
+
+func (r DefaultCompactReader) ReadString(fieldName string) *string {
+	fd := r.getFieldDefinitionChecked(fieldName, pserialization.FieldKindString)
+	value := r.getVariableSize(fd, func(in *ObjectDataInput) interface{} {
+		str := in.ReadString()
+		return &str
+	})
+	if value == nil {
+		return nil
+	}
+	return value.(*string)
+}
+
+func NewDefaultCompactReader(serializer CompactStreamSerializer, input *ObjectDataInput, schema Schema) DefaultCompactReader {
+	numberOfVarSizeFields := schema.numberOfVarSizeFields
+	var variableOffsetsPosition, dataStartPosition, finalPosition int32
+	if numberOfVarSizeFields == 0 {
+		dataStartPosition = input.Position()
+		finalPosition = dataStartPosition + schema.fixedSizeFieldsLength
+		variableOffsetsPosition = 0
+	} else {
+		dataLength := input.readInt32()
+		dataStartPosition = input.Position()
+		variableOffsetsPosition = dataStartPosition + dataLength
+		finalPosition = variableOffsetsPosition + numberOfVarSizeFields
+	}
+	input.SetPosition(finalPosition)
+	return DefaultCompactReader{
+		schema:                  schema,
+		in:                      input,
+		serializer:              serializer,
+		dataStartPosition:       dataStartPosition,
+		offsetReader:            &ByteOffsetReader{},
+		variableOffsetsPosition: variableOffsetsPosition,
+	}
+}
+
+func (r *DefaultCompactReader) getFieldDefinition(fieldName string) FieldDescriptor {
+	fd := r.schema.GetField(fieldName)
+	if fd == nil {
+		panic(r.unknownField(fieldName))
+	}
+	return *fd
+}
+
+func (r *DefaultCompactReader) getFieldDefinitionChecked(fieldName string, fieldKind pserialization.FieldKind) FieldDescriptor {
+	fd := r.getFieldDefinition(fieldName)
+	if fd.fieldKind != fieldKind {
+		panic(r.unexpectedFieldKind(fd.fieldKind, fieldName))
+	}
+	return fd
+}
+
+func (r *DefaultCompactReader) readFixedSizePosition(fd FieldDescriptor) int32 {
+	primitiveOffset := fd.offset
+	return primitiveOffset + r.dataStartPosition
+}
+
+func (r *DefaultCompactReader) unknownField(fieldName string) error {
+	return ihzerrors.NewSerializationError(fmt.Sprintf("Unknown field name '%s' for %s", fieldName, r.schema.String()), nil)
```

Comment:
```suggestion
	return ihzerrors.NewSerializationError(fmt.Sprintf("Unknown field name '%s' for %s", fieldName, r.schema), nil)
```
redundant

Reply:
`%s` identifier instructs it to call `func (T)String() string` of the argument if it implements the `Stringer` interface. So it is okay to call '%s'

Reply:
https://go.dev/play/p/Lv7QAnZIkBA

Reply:
Ah ok but it should not be pointer receiver, a value receiver otherwise I get an error: 
![image](https://user-images.githubusercontent.com/32355782/174052392-cb2c1141-3a67-4c0f-8672-3dbd2a263037.png)